File: Number of Substrings Containing All Three Characters.py
# https://leetcode.com/problems/number-of-substrings-containing-all-three-characters/?envType=daily-question&envId=2025-03-11
# Number of Substrings Containing All Three Characters

class Solution:
    def numberOfSubstrings(self, s: str) -> int:
        
        n = len(s)
        count = [0, 0, 0]
        l = 0
        res = 0
        
        for r in range(n):
            count[ord(s[r]) - ord('a')] += 1
            
            while count[0] > 0 and count[1] > 0 and count[2] > 0:
                res += (n - r)
                count[ord(s[l]) - ord('a')] -= 1
                l += 1
        
        return res

File: Number of Substrings Containing All Three Characters.js
// https://leetcode.com/problems/number-of-substrings-containing-all-three-characters/?envType=daily-question&envId=2026-06-30
// Number of Substrings Containing All Three Characters

/**
 * @param {string} s
 * @return {number}
 */
var numberOfSubstrings = function(s) {
    
    const n = s.length;
    const count = [0, 0, 0];
    let left = 0;
    let res = 0;

    for (let right = 0; right < n; right++) {
        count[s.charCodeAt(right) - 'a'.charCodeAt(0)]++;

        while (count[0] > 0 && count[1] > 0 && count[2] > 0) {
            res += (n - right);
            count[s.charCodeAt(left) - 'a'.charCodeAt(0)]--;
            left++;
        }
    }

    return res;

};